blog/2021-11-19-pwa-tech-1/index.md

```diff
@@ -0,0 +1,272 @@
+
+Putting these things together, here’s what happens in a scenario where the kill switch is used:
+
+1. Previously, an app uses a service worker named `service-worker.js` that caches the app’s assets and serves them from the cache without accessing the network.
+2. A new version of the app that doesn’t have PWA enabled gets deployed to the server, but since it’s a platform app, it still includes a `service-worker.js` file.
+3. When a user visits the app in the browser, the existing service worker loads the app from the cached files – this is the old version of the app (an app *without* `service-worker.js` would get stuck here). Then, the browser checks the server for a file called `service-worker.js`. It finds one, and finds that it’s byte-different from the currently installed service worker, so it installs this new one.
+4. When the new service worker script is run, it sees that PWA is not enabled, so it activates the kill switch mode. When it installs, it skips waiting to activate and take control immediately, then unregisters itself, clears the caches, and reloads the page.
+5. When the page reloads, the app files are requested from the server, which responds with the new ones, and the newly-deployed app loads normally.
+
+By including this kill-switch mode, we prevent apps from getting stuck in the future *and* we unstick apps that have been stuck in the past, like the core apps mentioned above.
+
+## Conclusion
+
+This concludes the discussion of the foundational PWA features that have been introduced in the App Platform, including installability, build tooling to read an app’s config and compile a service worker, caching strategies, and service worker updates and lifecycle management. Hopefully this closer look is interesting for its technical design and is helpful for a deeper understanding of how these features work together to enable offline capability that you’ll start to see in DHIS2 apps, starting with the Dashboard app in version 2.37.0.
+
+A following article will describe design challenges and solutions for creating the “cacheable sections” and some other App Runtime features that were described in the [PWA introduction article](https://developers.dhis2.org/blog/2021/11/introducing-pwa) released last week; look out for that one if you found this one interesting!
```

Comment:
When linking to internal pages, always refer to the `.md` file to allow Docusaurus to resolve the canonical link and detect broken links at build time.

As a bonus it works better locally, and, on GitHub.